### PR TITLE
[WIP] UNet VAE

### PR DESCRIPTION
**Pull request by annikabrundyn**
TODOs:

- [x] add image logging + saving
- [ ] generalize to use any encoder / decoder 
- [ ] generalize to work with multiple frames

### CODE REVIEW

File: src/models/right_view/vae/vae_unet.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        # variational
+        mu = self.fc_mu(emb)
+        logvar = self.fc_logvar(emb)
+
+        # kl
+        std = torch.exp(logvar / 2)
+        P = torch.distributions.Normal(torch.zeros_like(mu), torch.ones_like(std))
+        Q = torch.distributions.Normal(mu, std)
+        z = Q.rsample()
+        kl = (Q.log_prob(z) - P.log_prob(z)).sum(-1)
```
Comment (jeswan) [inline src/models/right_view/vae/vae_unet.py:81]: might be cleaner if we split this into another function
```
  def reparameterize(self, mu, logvar):
    if self.training:
      std = torch.exp(0.5*logvar)
      eps = torch.randn_like(std)
      z = mu + std*eps
    else:
      z = mu
    return z
```

File: src/models/right_view/vae/vae_unet.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+        # project emb and z to match original decoder dims
+        first_dec_out = xi[-1]
+        z = self.projection_1(z)
+        z = z.view(first_dec_out.size())
```
Comment (jeswan) [inline src/models/right_view/vae/vae_unet.py:86]: might be cleaner with we split this into a different function:
```
 def reparameterize(self, mu, logvar):
    if self.training:
      std = torch.exp(0.5*logvar)
      eps = torch.randn_like(std)
      z = mu + std*eps
    else:
      z = mu
    return z
```

File: src/models/right_view/vae/vae_unet.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+        # variational
+        mu = self.fc_mu(emb)
+        logvar = self.fc_logvar(emb)
+
+        # kl
+        std = torch.exp(logvar / 2)
+        P = torch.distributions.Normal(torch.zeros_like(mu), torch.ones_like(std))
+        Q = torch.distributions.Normal(mu, std)
+        z = Q.rsample()
```
Comment (jeswan) [inline src/models/right_view/vae/vae_unet.py]: Does this result in reparameterization (see code snippet below)?

Comment (jeswan) [reply src/models/right_view/vae/vae_unet.py]: Ah, looks like it does

File: src/models/right_view/vae/vae_unet_model.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+                                   bilinear=bilinear)
+
+    def forward(self, x):
+        return self.net(x)
+
+    def step(self, batch, batch_idx):
+        img, target = batch
+        pred, kl, std_min, std_max = self(img)
+
+        mse_loss = ((pred - target) ** 2).mean(dim=(1, 2, 3))
```
Comment (jeswan) [inline src/models/right_view/vae/vae_unet_model.py:55]: why were changes necessary here?

File: src/models/right_view/vae/vae_unet.py (hunk 1 above)
Comment (jeswan) [inline src/models/right_view/vae/vae_unet.py]: alternative way to calculate KLD. Also, can you point me to the equation that was used for P and Q?
```
  KLD = -0.5 * torch.sum(1 + logvar - mu.pow(2) - logvar.exp(), 1).mean()
```

File: src/models/right_view/vae/vae_unet.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+        # embedding
+        emb = xi[-1]
+        emb = emb.view(emb.size(0), -1)
+        emb = self.fc(emb)
+
+        # variational
+        mu = self.fc_mu(emb)
+        logvar = self.fc_logvar(emb)
+
+        # kl
```
Comment (jeswan) [inline src/models/right_view/vae/vae_unet.py:79]: I think we should split this into another function so we can change functionality during eval. Something like:
```
  def reparameterize(self, mu, logvar):
    if self.training:
      std = torch.exp(0.5*logvar)
      eps = torch.randn_like(std)
      z = mu + std*eps
    else:
      z = mu
    return z
```

File: src/models/right_view/vae/vae_unet_model.py (hunk 4 above)
Comment (annikabrundyn) [reply src/models/right_view/vae/vae_unet_model.py:55]: basically we want to sum mse + kl for every element in the batch first and then average. we don't want (average mse for all samples in batch) + (average kl for all samples in the batch)

does that make sense?

File: src/models/right_view/vae/vae_unet.py (hunk 1 above)
Comment (annikabrundyn) [reply src/models/right_view/vae/vae_unet.py]: I've swapped out for analytical form now

File: src/models/right_view/vae/vae_unet.py (hunk 5 above)
Comment (annikabrundyn) [reply src/models/right_view/vae/vae_unet.py:79]: added this too

**Pull request merged**